core/example/benchmark/Cabana_CommPerformance.cpp

```diff
@@ -0,0 +1,604 @@
+    void start( const int data_point )
+    {
+        if ( !_is_stopped[data_point] )
+            throw std::logic_error( "attempted to start a running timer" );
+        auto now = std::chrono::high_resolution_clock::now();
+        _starts[data_point].push_back( now );
+        _is_stopped[data_point] = false;
+    }
+
+    // Stop the timer at the given data point. It is assumed that this
+    // function is called collectively across all ranks in the communicator.
+    void stop( const int data_point )
+    {
+        if ( _is_stopped[data_point] )
+            throw std::logic_error( "attempted to stop a stopped timer" );
+        auto now = std::chrono::high_resolution_clock::now();
+        _data[data_point].push_back( std::chrono::duration_cast<std::chrono::microseconds>(
+                                         now - _starts[data_point].back() ) );
+        _is_stopped[data_point] = true;
+    }
+
+    // Get the number of timer data points.
+    int numDataPoint() const
+    {
+        return _data.size();
+    }
+
+    // Get the number of timer runs for a given data point.
+    int numRun( const int data_point ) const
+    {
+        return _data[data_point].size();
+    }
+
+    // Write timer results on rank 0. Provide the data points so they can be
+    // injected into the table. This function does collective communication.
+    template<typename Scalar>
+    void outputResults( std::ostream& stream,
+                        const std::string& data_point_name,
+                        const std::vector<Scalar>& data_point_vals ) const
+    {
+        // Get comm rank;
+        int comm_rank;
+        MPI_Comm_rank( _comm, &comm_rank );
+
+        // Get comm size;
+        int comm_size;
+        MPI_Comm_size( _comm, &comm_size );
+
+        // Write the data header.
+        if ( 0 == comm_rank )
+        {
+            stream << "\n";
+            stream << _name << "\n";
+            stream << "num_rank " << data_point_name << " min max ave" << "\n";
+        }
+
+        // Write out each data point
+        for ( std::size_t n = 0; n < _data.size(); ++n )
+        {
+            if ( !_is_stopped[n] )
+                throw std::logic_error( "attempted to output from a running timer" );
+
+            // Compute the minimum.
+            int local_min = (int) std::min_element(
+                _data[n].begin(), _data[n].end() )->count();
+            int global_min = 0.0;
+            MPI_Reduce( &local_min, &global_min, 1, MPI_INT, MPI_MIN, 0, _comm );
+
+            // Compute the maximum.
+            int local_max = (int) std::max_element(
+                _data[n].begin(), _data[n].end() )->count();
+            int global_max = 0.0;
+            MPI_Reduce( &local_max, &global_max, 1, MPI_INT, MPI_MAX, 0, _comm );
+
+            // Compute the average.
+            double local_sum = 0.0;
+            for ( auto d : _data[n] )
+                local_sum += (double) d.count();
+            double average = 0.0;
```

Comment:
Not changing this one to `int`?

Reply:
I guess here I was thinking an average could/should be represented as floating point because fractions of seconds are possible.

Reply:
To be clear, I would have done everything floating point durations all the way.  See my previous comment.

Reply:
OK guess I was confused there. I wanted my output units to be in microseconds. Would you have just done the duration cast to microseconds in the implementation but then stored in the class data as double?

Reply:
fixed per phone discussion